format/experimental/computeir/Expression.fbs

```diff
@@ -0,0 +1,351 @@
+  Following,
+  CurrentRow,
+}
+
+/// Boundary of a window
+table Bound {
+  impl: BoundImpl (required);
+}
+
+/// The kind of window function to be executed.
+enum Frame : uint8 {
+  Rows,
+  Range,
+}
+
+/// An expression representing a window function call.
+table WindowCall {
+  /// The kind of window frame
+  kind: Frame;
+  /// The expression to operate over
+  expression: Expression (required);
+  /// Partition keys
+  partitions: [Expression] (required);
+  /// Sort keys
+  orderings: [SortKey] (required);
+  /// Lower window bound
+  lower_bound: Bound (required);
+  /// Upper window bound
+  upper_bound: Bound (required);
+}
+
+/// A canonical (probably SQL equivalent) function
+enum CanonicalAggregateId : uint32 {
+  All,
+  Any,
+  Count,
+  CountTable,
+  Mean,
+  Min,
+  Max,
+  Product,
+  Sum,
+  Variance,
+  StandardDev,
+}
+
+
+table CanonicalAggregate {
+  id: CanonicalAggregateId;
+}
+
+table NonCanonicalAggregate {
+  name_space: string;
+  name: string (required);
+}
+
+union AggregateImpl {
+  CanonicalAggregate,
+  NonCanonicalAggregate,
+}
+
+table AggregateCall {
+  /// The kind of aggregate function being executed
+  kind: AggregateImpl (required);
+
+  /// Aggregate expression arguments
+  arguments: [Expression] (required);
+
+  /// Possible ordering.
+  orderings: [SortKey];
+
+  /// optional per-aggregate filtering
+  predicate: Expression;
+}
+
+/// An expression is one of
+/// - a Literal datum
+/// - a reference to a field from a Relation
+/// - a call to a named function
+/// - a case expression
+/// - a cast expression
+/// - an extract operation
+/// - a window function call
+/// - an aggregate function call
+///
+/// The expressions here that look like function calls such as
+/// Cast,Case and Extract are special in that while they might
+/// fit into a Call, they don't cleanly do so without having
+/// to pass around non-expression arguments as metadata.
+///
+/// AggregateCall and WindowCall are also separate variants
+/// due to special options for each that don't apply to generic
+/// function calls. Again this is done to make it easier
+/// for consumers to deal with the structure of the operation
+union ExpressionImpl {
+  Literal,
+  FieldRef,
+  Call,
+  Case,
+  Cast,
+  Extract,
+  WindowCall,
+  AggregateCall,
+}
+
+/// Expression types
+///
+/// Expressions have a concrete `impl` value, which is a specific operation
+/// They also have a `type` field, which is the output type of the expression,
+/// regardless of operation type.
+///
+/// The only exception so far is Cast, which has a type as input argument, which
+/// is equal to output type.
+table Expression {
+  impl: ExpressionImpl (required);
+
+  /// The type of the expression.
+  ///
+  /// This is a field, because the Type union in Schema.fbs
+  /// isn't self-contained: Fields are necessary to describe complex types
+  /// and there's currently no reason to optimize the storage of this.
+  type: org.apache.arrow.flatbuf.Field;
```

Comment:
I think the purpose of this field is potentially being misunderstood.

There are two very broad use cases that this is designed for:

1. A producer and a consumer that are not self-contained. This is the broadest use case and I think will be the primary way in which the IR is used. In this scenario, the producer inserts a type into this field as the output of its type system. The consumer is then required to adhere to what the producer asked for, or return an error.
2. A producer and consumer that are self-contained. This is for a system such as a relational database (e.g., DuckDB), where the IR is being used by a system that controls both the producer and the consumer. The purpose of the type field here is to have a place for the output of a type derivation step to be stored after its derivation for later consumption.

Additionally, the field is optional to allow for the type derivation phase to happen at any point in time.

What I think would be a mistake is to not have this field here and assume the producer and consumer will necessarily derive the same output type for every expression. We've already said that defining output type derivation rules are the responsibility of producers. Are we changing that decision?

Regarding performance, I don' t think that's something we should focus on until we're happy with the design that supports the use cases we want to support, and until we have some clarity on what is actually expensive in real-world use cases.

Reply:
I think that the rational is inverted: usually planners (IR producers) are the ones that need to know which signatures the consumers accept, so that they can plan eventual detours (e.g. perform some casts to match signatures).

For example, if a consumer does not support `(add,Date32,Duration(seconds))`, there are ways around this via casting and arithmetics; the planner may want to get around this limitation.

Some examples where typing is needed:

```
sum(f32) -> (f32 or f64)?
sum(i32) -> (i32 or i64)?
add(i32,i32) -> (i32 or i64)?
```

different code will need to run depending on the requested output type. If an IR consumer does not support some of the calls, the producer needs to know that so that it can emit a different set of operators (e.g. perform some casts before and after the main operator).

`DataFusion` dynamic dispatch is based on this idea; the physical planner plans out which physical expressions will be used based on the input schema. Broadly, it goes like this:

1. every operator has an ordered set of valid input types that it supports (e.g. `add` could be `(i32,i32),(i64,i64),(f32,f32),(f64,f64)` in this order; ordering is from least informative (in Shannon's notion) to most informative.
2. for every operator, there is a map between input and output: `output_type(op,arg_types) -> arg_type` (e.g. `output_type("add", [i32,i32]) -> i64`)
3. the first lossless coercion to fit the signature is selected (e.g. if input types are `(i32,i64)`, they are mapped to `(i64,i64)` prior to being passed to `add` (since the `(i32,i32)` would be lossy)).

Given an input schema (vector of of `DataType`), all IR calls are known in advance via the type resolution of the physical nodes available + applied coercion.  `DataFusion`'s planner knows in advance which physical nodes are implemented and uses this information to plan out the operations (since it is in the same source code). UDFs and UDAFs in DataFusion must register this information; in fact, UDFs and UDAFs use the same signatures as builtin functions; the only way they are special is that their implementation is not known to DataFusion and thus can't be inlined (which is fine).

I would imagine that consumers would have a registry of implemented functions, which they would inform producers about _before planning_, so that the producer can error out or "dodge via cast" _during planning_. The producer would use the registry to decide which physical plan it would pass down to the consumer, pretty much like an architecture that declares the registry in their user manual and that compilers use.

The main question for me is whether we want to use logical types or physical types. E.g. is it relevant that a consumer does not implement `add(date32,duration(seconds))` if it implements `add(int32,int64)`? My understanding is that it should not matter since logical semantics can be handled by an IR planner.

Reply:
Please take this with the usual caveat that I am no expert in this field.

> A producer and a consumer that are not self-contained. This is the broadest use case and I think will be the primary way in which the IR is used. In this scenario, the producer inserts a type into this field as the output of its type system. The consumer is then required to adhere to what the producer asked for, or return an error.

This sounds like the user needs to create overloads based solely on the return type.  I feel like that will be rare enough that it can be handled via the name or the call arguments/options.  For example, `add_unchecked(int32, int32) => int32` and `add_or_widen(int32, int32) => int64`.  It would be odd for me to pick which behavior I get by specifying the return type.

> A producer and consumer that are self-contained. This is for a system such as a relational database (e.g., DuckDB), where the IR is being used by a system that controls both the producer and the consumer. The purpose of the type field here is to have a place for the output of a type derivation step to be stored after its derivation for later consumption.

This feels like a slippery slope.  Wouldn't it be better for DuckDB in this case to extend the fbs with their own fields?

Reply:
> I think that the rational is inverted: usually planners (IR producers) are the ones that need to know which signatures the consumers accept, so that they can plan eventual detours (e.g. perform some casts to match signatures).

I get what you're saying here, but right now the IR is explicitly not concerned with how (or even whether) a producer is aware of what types of operations are accepted by a consumer.

This is compatible with systems that *do* have additional knowledge; the IR is designed to be unopinionated about how a producer deduces types.

The flow of execution you're describing is compatible with the current IR if I understand the flow correctly: the producer can query the possible operations, figure out how (and whether) to execute the desired operations, and then put all of that into IR and send it to the consumer.

> I would imagine that consumers would have a registry of implemented functions, which they would inform producers about before planning, so that the producer can error out or "dodge via cast" during planning.

Indeed, see the discussion about registries/signatures here https://github.com/apache/arrow/pull/10934/files#r697016776

> The main question for me is whether we want to use logical types or physical types.

You can use either, unless the type systems for representing physical and logical types are distinct trees. AFAICT Arrow doesn't distinguish between logical and physical types in `Schema.fbs`. Is that correct?

Reply:
@westonpace 

> This sounds like the user needs to create overloads based solely on the return type. 

Assuming by "user" you mean IR producer, then what you said isn't correct if I understand you correctly. The IR producer's behavior with respect to what IR it chooses to generate is entirely up to the producer. The IR has nothing to do with type derivation at all.

> This feels like a slippery slope. Wouldn't it be better for DuckDB in this case to extend the fbs with their own fields?
Maybe. We can certainly just leave out types altogether and tell consumers that they need to have a way to determine the type of an expression in IR. To me that seems like it's going to breed a bunch of code duplication across different consumers, but maybe it's less onerous than I think.

Reply:
> Assuming by "user" you mean IR producer, then what you said isn't correct if I understand you correctly. The IR producer's behavior with respect to what IR it chooses to generate is entirely up to the producer. The IR has nothing to do with type derivation at all.

Correct, but can you give a concrete example of when an IR producer would need to specify the output type?  Why wouldn't a cast be sufficient?